File: src/parser/sema.c
#ifndef SEMA_H
#define SEMA_H

#include "../std/Allocator.c"
#include "../std/eql.c"
#include "parser.c"
#include <stdbool.h>

typedef struct {
  Slice(char) name;
  bool read;
  bool constant;
} Binding;

DefSlice(Binding);
DefVec(Binding);
DefResult(Vec_Binding);

static bool nameListHasString(Slice(Binding) list, Slice(char) string) {
  for (size_t i = 0; i < list.len; i++) {
    if (eql(list.ptr[i].name, string))
      return true;
  }
  return false;
}

static void analyzeStatement(Vec(Binding) * names, Statement stmt);

static void analyzeExpression(Allocator ally, Slice(Binding) names,
                              Expression expr) {
  switch (expr.type) {
  case IdentifierExpression: {
    if (!nameListHasString(names, expr.identifier)) {
      if (!eql(expr.identifier, (Slice_char){.ptr = "print", .len = 5})) {
        fprintf(stdout, "Referring to undeclared name: %1.*s\n",
                (int)expr.identifier.len, expr.identifier.ptr);
      }
    } else {
      for (size_t i = 0; i < names.len; i++) {
        if (eql(names.ptr[i].name, expr.identifier)) {
          names.ptr[i].read = true;
        }
      }
    }
  } break;
  case CallExpression: {
    analyzeExpression(ally, names, *expr.call.callee);
    for (size_t i = 0; i < expr.call.parameters_len; i++) {
      analyzeExpression(ally, names, expr.call.parameters[i]);
    }
  } break;
  case ArithmeticExpression: {
    analyzeExpression(ally, names, *expr.arithmetic.left);
    analyzeExpression(ally, names, *expr.arithmetic.right);
  } break;
  case FunctionExpression: {
    Result(Vec_Binding) locals_res = createVec(ally, Binding, 1);
    if (!locals_res.ok)
      panic(locals_res.err);
    Vec(Binding) locals = locals_res.val;
    for (size_t i = 0; i < expr.function_expression.parameters_len; i++) {
      Expression param = expr.function_expression.parameters[i];
      Binding binding = {
          .name = param.identifier,
          .constant = false,
          .read = false,
      };
      if (!append(&locals, Binding, &binding)) {
        panic("Failed to append to names");
      }
    }
    analyzeStatement(&locals, *expr.function_expression.body);
  } break;
  case NumericExpression:
  case StringExpression: {
    // no-op
  }
  }
}

static void analyzeStatement(Vec(Binding) * names, Statement stmt) {
  switch (stmt.type) {
  case DeclarationStatement: {
    if (nameListHasString(names->slice, stmt.declaration.name)) {
      fprintf(stdout, "Double declaration of: %1.*s\n",
              (int)stmt.declaration.name.len, stmt.declaration.name.ptr);
      return;
    }
    analyzeExpression(names->ally, names->slice, stmt.declaration.value);
    Binding binding = {.name = stmt.declaration.name,
                       .constant = stmt.declaration.constant,
                       .read = false};
    if (!append(names, Binding, &binding)) {
      panic("analyze: Failed to append to names");
    }
  } break;
  case AssignmentStatement: {
    if (!nameListHasString(names->slice, stmt.assignment.name)) {
      fprintf(stdout, "Assignment to undeclared name: %1.*s\n",
              (int)stmt.assignment.name.len, stmt.assignment.name.ptr);
    } else {
      for (size_t j = 0; j < names->slice.len; j++) {
        if (eql(names->slice.ptr[j].name, stmt.assignment.name)) {
          if (names->slice.ptr[j].constant) {
            fprintf(stdout, "Assignment to constant: %1.*s\n",
                    (int)stmt.assignment.name.len, stmt.assignment.name.ptr);
          }
        }
      }
    }
    analyzeExpression(names->ally, names->slice, stmt.assignment.value);
  } break;
  case ExpressionStatement: {
    analyzeExpression(names->ally, names->slice, stmt.expression);
  } break;
  case IfStatement: {
    analyzeExpression(names->ally, names->slice, stmt.if_statement->condition);
    analyzeStatement(names, *stmt.if_statement->consequence);
    if (stmt.if_statement->alternate) {
      analyzeStatement(names, *stmt.if_statement->alternate);
    }
  } break;
  case WhileStatement: {
    analyzeExpression(names->ally, names->slice,
                      stmt.while_statement->condition);
    analyzeStatement(names, *stmt.while_statement->body);
  } break;
  case BlockStatement: {
    for (size_t i = 0; i < stmt.block->statements.len; i++) {
      analyzeStatement(names, stmt.block->statements.ptr[i]);
    }
  } break;
  case ReturnStatement: {
    analyzeExpression(names->ally, names->slice, *stmt.return_statement);
  } break;
  case InlineBatchStatement: {
  } break;
  case StatementEOF: {
    panic("StatementEOF");
  } break;
  }
}

static void analyze(Allocator ally, Program prog) {
  Result(Vec_Binding) names_res = createVec(ally, Binding, 4);
  if (!names_res.ok)
    panic(names_res.err);
  Vec(Binding) names = names_res.val;
  for (size_t i = 0; i < prog.statements.len; i++) {
    Statement stmt = prog.statements.ptr[i];
    analyzeStatement(&names, stmt);
  }
  for (size_t i = 0; i < names.slice.len; i++) {
    Binding b = names.slice.ptr[i];
    if (!b.read) {
      fprintf(stdout, "Unused %s: %1.*s\n",
              b.constant ? "constant" : "variable", (int)b.name.len,
              b.name.ptr);
    }
  }
}
#endif /* SEMA_H */

File: src/parser/sema.zig
const std = @import("std");
const Slice = @import("../std/Slice.zig").Slice;
const vec = @import("../std/Vec.zig");
const alloc = @import("../std/Allocator.zig");
const Allocator = alloc.Allocator;
const p = @import("parser.zig");

pub const Binding = struct {
    name: []const u8,
    read: bool,
    constant: bool,
};

pub fn nameListHasString(bindings: []const Binding, string: []const u8) bool {
    for (bindings) |binding| {
        if (std.mem.eql(u8, binding.name, string)) return true;
    }
    return false;
}

pub fn analyzeExpression(ally: std.mem.Allocator, names: []Binding, expr: p.Expression) std.mem.Allocator.Error!void {
    const stdout = std.io.getStdOut().writer();
    switch (expr.tag) {
        .identifier => {
            if (!nameListHasString(names, expr.x.identifier)) {
                if (!std.mem.eql(u8, expr.x.identifier, "print")) {
                    stdout.print(
                        "Referring to undeclared name: {s}\n",
                        .{expr.x.identifier},
                    ) catch {};
                }
            } else for (names) |*binding|
                if (std.mem.eql(u8, binding.name, expr.x.identifier)) {
                    binding.read = true;
                };
        },
        .call => {
            try analyzeExpression(ally, names, expr.x.call.callee.*);
            for (expr.x.call.parameters[0..expr.x.call.parameters_len]) |param|
                try analyzeExpression(ally, names, param);
        },
        .arithmetic => {
            try analyzeExpression(ally, names, expr.x.arithmetic.left.*);
            try analyzeExpression(ally, names, expr.x.arithmetic.right.*);
        },
        .function => {
            var locals = try std.ArrayList(Binding).initCapacity(ally, 1);
            for (expr.x.function_expression.parameters[0..expr.x.function_expression.parameters_len]) |param| {
                const binding = Binding{
                    .name = param.x.identifier,
                    .constant = false,
                    .read = false,
                };
                try locals.append(binding);
            }
            try analyzeStatement(&locals, expr.x.function_expression.body.*);
        },
        .numeric, .string => {},
    }
}

pub fn analyzeStatement(names: *std.ArrayList(Binding), stmt: p.Statement) !void {
    const stdout = std.io.getStdOut().writer();
    switch (stmt.tag) {
        .declaration => {
            if (nameListHasString(names.items, stmt.x.declaration.name)) {
                stdout.print(
                    "Double declaration of: {s}\n",
                    .{stmt.x.declaration.name},
                ) catch {};
                return;
            }
            try analyzeExpression(names.allocator, names.items, stmt.x.declaration.value);
            const binding = Binding{
                .name = stmt.x.declaration.name,
                .constant = stmt.x.declaration.constant,
                .read = false,
            };
            try names.append(binding);
        },
        .assignment => {
            if (!nameListHasString(names.items, stmt.x.assignment.name)) {
                stdout.print(
                    "Assignment to undeclared name: {s}\n",
                    .{stmt.x.assignment.name},
                ) catch {};
            } else {
                for (names.items) |binding| {
                    if (std.mem.eql(u8, binding.name, stmt.x.assignment.name)) {
                        if (binding.constant) {
                            stdout.print(
                                "Assignment to constant: {s}\n",
                                .{stmt.x.assignment.name},
                            ) catch {};
                        }
                    }
                }
            }
            try analyzeExpression(names.allocator, names.items, stmt.x.assignment.value);
        },
        .expression => {
            try analyzeExpression(names.allocator, names.items, stmt.x.expression);
        },
        .@"if" => {
            try analyzeExpression(names.allocator, names.items, stmt.x.@"if".condition);
            try analyzeStatement(names, stmt.x.@"if".consequence.*);
            if (stmt.x.@"if".alternate) |alt| try analyzeStatement(names, alt.*);
        },
        .@"while" => {
            try analyzeExpression(names.allocator, names.items, stmt.x.@"while".condition);
            try analyzeStatement(names, stmt.x.@"while".body.*);
        },
        .block => {
            for (stmt.x.block.statements) |s| try analyzeStatement(names, s);
        },
        .@"return" => if (stmt.x.@"return") |ret|
            try analyzeExpression(names.allocator, names.items, ret.*),
        .inline_batch => {},
        .eof => @panic("StatementEOF"),
    }
}

pub fn analyze(ally: std.mem.Allocator, prog: p.Program) !void {
    const stdout = std.io.getStdOut().writer();
    var names = try std.ArrayList(Binding).initCapacity(ally, 4);
    for (prog.statements) |stmt| {
        try analyzeStatement(&names, stmt);
    }
    for (names.items) |b| {
        if (!b.read) {
            stdout.print("Unused {s}: {s}\n", .{
                if (b.constant) "constant" else "variable",
                b.name,
            }) catch {};
        }
    }
}
